kvserver: prevent race in Replica creation

pav-kv opened the pull request:
Previously, `getOrCreateReplica`:
- Creates a half-baked / "unloaded" `Replica` with `newUnloadedReplica`
- Inserts it to `replicasByRangeID`
- Leaves only the `Replica`'s locks acquired
- Finalizes / "loads" the `Replica`

This leaves a window of time during which the not properly created `Replica` is accessible via `Store` (in a limited way though, since its locks are still held). This makes space for a number of potential bugs.

After this commit:
- An uninitialized `Replica` is fully created
- Then it's inserted to the replicas map
- This is done with no regression in performance (in fact, there is a bonus)

As a result, the `Replica` is never accessible via `Store` until it's properly created.

Bonuses:
- `RangeTombstone` is read only once per replica creation (previously twice).
- The concept of unloaded replicas is now easily removable, because the replica creation code is consolidated, and is not interleaved with `Store` operations.

Informs #94911
Release note: none

pav-kv committed:
kvserver: prevent race in Replica creation

Previously, getOrCreateReplica:
- Creates a half-baked / "unloaded" Replica with newUnloadedReplica
- Inserts it to replicasByRangeID
- Leaves only the Replica's locks acuired
- Finalizes / "loads" the Replica

This leaves a window of time during which the not properly created Replica is
accessible via Store (in a limited way though, since its lockes are still
held). This makes space for a number of potential bugs.

After this commit:
- An uninitialized Replica is fully created
- Then it's inserted to the replicas map
- This is done with no regression in performance (in fact, there is a bonus)

As a result, the Replica is never accessible via Store until it's properly
created.

Bonuses:
- RangeTombstone is read only once per replica creation (previously twice).
- The concept of unloaded replicas is now easily removable, because the replica
  creation code is consolidated, and is not interleaved with Store operations.

Release note: none

pav-kv committed:
kvserver: factor out tryGetOrCreateReplica fast-path

Release note: none

pav-kv committed:
kvserver: unindent tryGetReplica

Release note: none

pav-kv committed:
kvserver: remove redundant replicaID != 0 check

ReplicaID has been checked to be non-0 up the stack. Also, it's ok to treat 0
as a sentinel value and return RaftGroupDeletedError (replica ID 0 is "always
deleted").

Release note: none

tbg reviewed (commented):
LGTM, just need to deal with the FIXME.

pav-kv committed:
kvserver: separate adding replicas to Store maps

In this commit addReplicaInternalLocked now requires the replica to be in the
replicasByRangeID map unconditionally, i.e. that addReplicaToRangeMapLocked has
been previously called for this replica.

This holds true in the production code because each Replica is either:
- created and initialized in Store.Start, in which case it's added to both maps
- created in getOrCreateReplica, and later initialized in a split trigger or
	snapshot application; in these cases the additions to these maps happen
	sequentially too

Release note: none

pavelkalinnikov commented on pkg/kv/kvserver/store_create_replica.go:
```diff
@@ -325,15 +323,31 @@ func (s *Store) tryGetOrCreateReplica(
+	}
+	s.mu.uninitReplicas[repl.RangeID] = repl
+	s.mu.Unlock()
+	// TODO(pavelkalinnikov): since we were holding s.mu anyway, consider
```
Agree that shouldn't over-optimize here, leaving as is for now.

pavelkalinnikov commented on pkg/kv/kvserver/store.go:
```diff
@@ -1893,7 +1893,12 @@ func (s *Store) Start(ctx context.Context, stopper *stop.Stopper) error {
 		s.mu.Lock()
-		err = s.addReplicaInternalLocked(rep)
+		// TODO(pavelkalinnikov): hide these in Store's replica create functions.
+		// TODO(pavelkalinnikov): addReplicaToRangeMapLocked needs a locked repl.mu.
```
(pre-existing)

pavelkalinnikov replied:
However, this might be ok since it's the "initialization" path. But it's not super-obvious. `addReplicaToRangeMapLocked` inserts the replica to the map **before** it accesses `repl.mu.quiescent`.

tbg replied:
Huh, I looked at `addReplicaToRangeMapLocked` and that it needs to look at the quiescing status seems fishy. Concretely, since replicas are always added to this map right when they're created (and are necessarily uninitialized - I'm ignoring the weirdo`(*Store).Start` here), how would they become unquiesced? An uninited replica physically *cannot* unquiesce:

https://github.com/cockroachdb/cockroach/blob/d40095138d8eff517f9126b71cbc7e9335ac0c30/pkg/kv/kvserver/replica_raft_quiesce.go#L95-L122

So the insertion into the quiescent there should be unconditional (and wouldn't need `r.mu` at all except maybe to assert).

For splits, I can only guess that the comment refers to

https://github.com/cockroachdb/cockroach/blob/38ad1f3cf9023329e84d1c0e961afc21bcf77c11/pkg/kv/kvserver/store_split.go#L355

but we already acquired the split lock, which already inserted the replica, so the second call [here](https://github.com/cockroachdb/cockroach/blob/0703807d018fb1c0f5352e8d83315b463c90d1e9/pkg/kv/kvserver/store_create_replica.go#L342) still has the replica in the map, and so there is no need to mess with the quiesced status. There's no such call on merges, so I'm hard-pressed to say what this is about. Possibly the comment predates the split/merge triggers in their entirety.

pavelkalinnikov replied:
> So the insertion into the quiescent there should be unconditional (and wouldn't need `r.mu` at all except maybe to assert).

I think this branch should rather be removed, it never happens (it inserts to ***un***quiesced).

Unquiescence can only happen on a `Replica` that's linked to `Store`, in reaction to some events/messages (right?). On top of that, as you say, unquiescence can only happen for initialized replicas. At the end state of this PR, `addToReplicasByRangeIDLocked` is called only for uninitialized replicas that aren't linked to `Store` and can't have got any messages.

This check was added in #27518, was it making more sense back then?

pavelkalinnikov replied:
Added another commit which removes this check. See the commit message with explanation.

tbg replied:
Thanks for digging up the PR that introduced this. This adds further weight that the check isn't useful: the PR's goal was to *avoid* adding uninited replicas as unquiesced. I assume the check was necessary because the call structure around replica lifecycles was even more circuitous then. :-)

pavelkalinnikov commented:
@tbg Dealt with the FIXME, see the last commit. I think it was safe even without the last commit.

tbg commented on pkg/kv/kvserver/store_create_replica.go:
```diff
@@ -361,17 +361,17 @@ func fromReplicaIsTooOldRLocked(toReplica *Replica, fromReplica *roachpb.Replica
+//
+// Returns an error if a different replica with the same range ID, or an
+// overlapping replica or placeholder exists in this Store.
 func (s *Store) addReplicaInternalLocked(repl *Replica) error {
```
It's because of https://github.com/cockroachdb/cockroach/issues/94912, the insertion happens in that method: https://github.com/cockroachdb/cockroach/blob/3d51c79d527871333bb690220da2d0ac90fc793d/pkg/kv/kvserver/store_create_replica.go#L448-L453

pavelkalinnikov replied:
Updated the commit message.

tbg commented on pkg/kv/kvserver/store_create_replica.go:
```diff
@@ -361,17 +361,17 @@ func fromReplicaIsTooOldRLocked(toReplica *Replica, fromReplica *roachpb.Replica
+//
+// Returns an error if a different replica with the same range ID, or an
+// overlapping replica or placeholder exists in this Store.
 func (s *Store) addReplicaInternalLocked(repl *Replica) error {
```
The method name is a bit nondescript, btw. How about `add[InitializedReplicas]ToReplicasByKey` with the `[]` part optional?

pavelkalinnikov replied:
Renamed to `addToReplicasByKeyLocked`. Also renamed the method inserting to `replicasByRangeID`, for consistency.

tbg reviewed (approved):
Liking what I see, I think we still have a bit to go until we reach a natural stopping point but this is great already, let's keep going.

pav-kv committed:
kvserver: rename method adding to replicasByKey

This method now only adds to replicasByKey, so the new name reflects this.

Release note: none

pav-kv committed:
kvserver: rename method adding to replicasByRangeID

The new name is consistent with a similar addToReplicasByKeyLocked method.

Release note: none

pav-kv committed:
kvserver: drop no-op quiescence check

After #73362, all uninitialized replicas are quiescent. The callers of
addToReplicasByRangeIDLocked are either inserting an unitialized Replica, or a
newly created initialized Replica which has not been linked to Store yet and
hence has not been unquiesced either. The quiescence check is thus a no-op, and
can be removed from addToReplicasByRangeIDLocked.

As a second "proof" for why it can be removed:

- rangeID is added to unquiescedReplicas in 2 places (Replica.maybeUnquiesce*),
  both of which also set r.mu.quiescent to false a few lines above.
- rangeID is removed from unquiescedReplicas in 2 places (Replica.quiesceLocked
	and Replica removal in Store.unlinkReplicaByRangeIDLocked). The first one
	sets r.mu.quiescent to true a few lines apart, and the second one outright
  removes the replica.

Thus, it is already true that r.mu.quiescent == false iff the replica's rangeID
is in unquiescedReplicas (with a little exception of when r is being removed
which does not impact our check).

Release note: none

pavelkalinnikov commented:
bors r=tbg

tbg commented:
Bors is off, just hit the button.

rail commented:
bors r-
We have temporarily disabled bors for merges. Please merge the PR manually when all required checks pass.
